pkg/compression/compression.go

```diff
@@ -69,26 +73,33 @@ func XzDecompressor(r io.Reader) (io.ReadCloser, error) {
 }
 
 // gzipCompressor is a CompressorFunc for the gzip compression algorithm.
-func gzipCompressor(r io.Writer, level *int) (io.WriteCloser, error) {
+func gzipCompressor(r io.Writer, metadata map[string]string, level *int) (io.WriteCloser, error) {
 	if level != nil {
 		return pgzip.NewWriterLevel(r, *level)
 	}
 	return pgzip.NewWriter(r), nil
 }
 
 // bzip2Compressor is a CompressorFunc for the bzip2 compression algorithm.
-func bzip2Compressor(r io.Writer, level *int) (io.WriteCloser, error) {
+func bzip2Compressor(r io.Writer, metadata map[string]string, level *int) (io.WriteCloser, error) {
 	return nil, fmt.Errorf("bzip2 compression not supported")
 }
 
 // xzCompressor is a CompressorFunc for the xz compression algorithm.
-func xzCompressor(r io.Writer, level *int) (io.WriteCloser, error) {
+func xzCompressor(r io.Writer, metadata map[string]string, level *int) (io.WriteCloser, error) {
 	return xz.NewWriter(r)
 }
 
 // CompressStream returns the compressor by its name
 func CompressStream(dest io.Writer, algo Algorithm, level *int) (io.WriteCloser, error) {
-	return internal.AlgorithmCompressor(algo)(dest, level)
+	m := map[string]string{}
+	return internal.AlgorithmCompressor(algo)(dest, m, level)
+}
+
+// CompressStreamWithMetadata returns the compressor by its name.  If the compression
+// generates any metadata, it is written to the provided metadata map.
```

Comment:
the metadata generated here will end up in an annotation.

I am fine to specify it, although `pkg/compression` doesn't seem the best place for it.  Do you have any suggestions where it should be documented?

Reply:
We don’t necessarily have to specify the _format_ here — we don’t carry a full specification of Zstd either.

We do need a “how would the caller ever use this?” so that there even can be a legitimate caller. Maybe something vaguely like:
> CompressStreamWithIntegrityData returns … . Compressing a stream will create integrity data that [is optional to  | must be] provided on decompression along with the compressed byte stream via $some_API, and that will allow the decompression to only use subsets of the compressed data (if the compressed data is seekable and most of the uncompressed data is already present via other means), _while still protecting integrity of the compressed stream against unwanted modification_. (In OCI container images, this metadata is usually carried in manifest annotations.)

That’s very likely not accurate, but it captures:
- How is the data used by the caller
- What is the benefit / consequence of not providing it
- Purpose/Applicability outside this narrow “compression” subpackage.

Reply:
@giuseppe Have  you fixed this comment?

Reply:
3 years later, this will be resolved in https://github.com/containers/image/pull/2605 .